File: ml-backend/app/db/models/wardrobe_item.py
"""
WardrobeItem ORM model.

Maps to the `wardrobe_items` PostgreSQL table.
Indexed fields (per architecture plan): user_id

Each row is one clothing item in a user's personal wardrobe.
The image_url points to the user's own photo stored in Backblaze B2
(the only images we store — stock and review images are URL-only).
"""

from __future__ import annotations

import uuid

from sqlalchemy import Index, String, Integer, ForeignKey, DateTime, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship
from sqlalchemy.sql import func

from app.db.database import Base


class WardrobeItem(Base):
    """
    A single item in a user's personal wardrobe.

    product_id is nullable: a user can add an item they didn't find on
    the platform (e.g., a gift or a foreign purchase).

    image_url is a Backblaze B2 URL — user-uploaded photos are the ONLY
    images we store in object storage. All other images are URL-only.

    image_embedding is stored in Pinecone for wardrobe gap analysis
    (finding which outfit categories the user is missing).
    """

    __tablename__ = "wardrobe_items"

    id: Mapped[str] = mapped_column(
        String(36), primary_key=True, default=lambda: str(uuid.uuid4())
    )
    user_id: Mapped[str] = mapped_column(
        String(36), ForeignKey("users.id", ondelete="CASCADE"), nullable=False
    )
    # Nullable: item may not be linked to a platform product
    product_id: Mapped[str] = mapped_column(
        String(36), ForeignKey("products.id", ondelete="SET NULL"), nullable=True
    )
    name: Mapped[str] = mapped_column(String(256), nullable=False)
    category: Mapped[str] = mapped_column(String(128), nullable=True)
    color: Mapped[str] = mapped_column(String(64), nullable=True)
    # Backblaze B2 URL — user uploaded this photo
    image_url: Mapped[str] = mapped_column(Text, nullable=True)
    purchase_price_inr: Mapped[int] = mapped_column(Integer, nullable=True)
    times_worn: Mapped[int] = mapped_column(Integer, nullable=False, default=0)
    added_at: Mapped[str] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), nullable=False
    )

    # Relationships
    user = relationship("User", back_populates="wardrobe_items")
    product = relationship("Product", back_populates="wardrobe_items")

    # ---------------------------------------------------------------------------
    # Indexes
    # ---------------------------------------------------------------------------

    __table_args__ = (
        # user_id: fetched on every wardrobe page load
        Index("ix_wardrobe_items_user_id", "user_id"),
    )

    def __repr__(self) -> str:
        return (
            f"<WardrobeItem id={self.id!r} user_id={self.user_id!r} "
            f"name={self.name!r} category={self.category!r}>"
        )

File: api-backend/src/models/WardrobeItem.js
/**
 * WardrobeItem Sequelize model — api-backend.
 *
 * Maps to the `wardrobe_items` table.
 * Indexed fields: user_id.
 *
 * product_id is nullable — items can be added without a platform product link.
 * image_url points to Backblaze B2 — user-uploaded wardrobe photos only.
 */

'use strict';

const { DataTypes } = require('sequelize');
const sequelize = require('../db/connection');

const WardrobeItem = sequelize.define(
  'WardrobeItem',
  {
    id: {
      type: DataTypes.UUID,
      defaultValue: DataTypes.UUIDV4,
      primaryKey: true,
    },
    user_id: {
      type: DataTypes.UUID,
      allowNull: false,
    },
    // Nullable: item may not be linked to a scraped product
    product_id: {
      type: DataTypes.UUID,
      allowNull: true,
    },
    name: {
      type: DataTypes.STRING(256),
      allowNull: false,
    },
    category: {
      type: DataTypes.STRING(128),
      allowNull: true,
    },
    color: {
      type: DataTypes.STRING(64),
      allowNull: true,
    },
    // Backblaze B2 URL — user uploaded this photo
    image_url: {
      type: DataTypes.TEXT,
      allowNull: true,
    },
    purchase_price_inr: {
      type: DataTypes.INTEGER,
      allowNull: true,
    },
    times_worn: {
      type: DataTypes.INTEGER,
      allowNull: false,
      defaultValue: 0,
    },
  },
  {
    tableName: 'wardrobe_items',
    timestamps: true,
    createdAt: 'added_at',
    updatedAt: false,
    indexes: [
      { fields: ['user_id'], name: 'ix_wardrobe_items_user_id' },
    ],
  }
);

module.exports = WardrobeItem;
